### ISSUE TITLE
Code signing with macOS Sandbox 

### ISSUE THREAD

**Issue by dagheyman** [pyinstaller/pyinstaller#2198]
I have a .app bundle built with pyinstaller.
When I `codesign` it without a sandbox it works fine, but when I add a Entitlements.plist file with `om.apple.security.app-sandbox` and some permissions the codesigning still works but the application fails with `Illegal instruction: 4`. 

Is this a known issue? Are bundles built with pyinstaller compatible with the sandbox?

**Comment by tallforasmurf** [comment]
No insult intended, but just to make sure, you have read the [mac app](http://pythonhosted.org/PyInstaller/usage.html#building-mac-os-x-app-bundles) part of the manual and the [codesign recipe](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-OSX-Code-Signing) that it links to?

Also I see this [stackoverflow question](http://stackoverflow.com/questions/14268887/what-is-the-illegal-instruction-4-error-and-why-does-mmacosx-version-min-10#14285922) which is probably relevant, esp. the last entry.

**Comment by dagheyman** [comment]
I have read those.

To reproduce with a minimal POC:

- PyInstaller 3.2 
- macOS Sierra

main.py:
`print("This is a poc.")`

Entitlements.plist:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>com.apple.security.app-sandbox</key>
    <true/>
</dict>
</plist>
```

**Create .app bundle:**
`pyinstaller main.py -w`

**Run program from bundle:**
```
./dist/main.app/Contents/MacOS/main
This is a poc.
```

**Sign with codesign:**
```
codesign --deep --sign [CERTIFICATE] dist/main.app/ --entitlements Entitlements.plist
```

**Run again:**
```
./dist/main.app/Contents/MacOS/main
Illegal instruction: 4
```

**Comment by skiptomyliu** [comment]
@dagheyman Have you made any progress?  I am encountering similar issues getting the sandbox code signing to work for store submission.  Attempting to embed my pyinstaller generated .app into my xcode cocoa project.

**Comment by dagheyman** [comment]
@skiptomyliu Unfortunately not, our project decided to move away from PyInstaller and build our app with Qt directly, using [PyOtherSide](https://github.com/thp/pyotherside) for the python parts.

**Comment by skiptomyliu** [comment]
Solved my related issue.  My issue is coming across the Illegal Instruction: 4 message is due to signing with child entitlements where I've set the com.apple.security.inherit  property to that of my parent (since I'm embedding the app in my osx app).  With the inherit, it is only allowed to run by the parent and not directly.

**Comment by davegradwell** [comment]
Further to the `Illegal instruction: 4` report, I see a bit more detail in my logs:
```
20/06/2017 20:16:47.711 com.apple.xpc.launchd[1]: (com.bonhardcomputing.deduplicator.674272[74402]) Service exited due to signal: Illegal instruction: 4
20/06/2017 20:16:47.725 com.apple.xpc.launchd[1]: (com.apple.ReportCrash[74404]) Endpoint has been activated through legacy launch(3) APIs. Please switch to XPC or bootstrap_check_in(): com.apple.ReportCrash
20/06/2017 20:16:47.733 sandboxd[142]: ([74403]) main(74403) deny forbidden-sandbox-reinit
20/06/2017 20:16:48.231 ReportCrash[74404]: Saved crash report for main[74403] version 2017.00.00 (2017.00.00) to /Users/dave/Library/Logs/DiagnosticReports/main_2017-06-20-201648_Daves-MacBook-Pro-2016.crash
20/06/2017 20:16:48.234 ReportCrash[74404]: Saved crash report for main[74402] version 2017.00.00 (2017.00.00) to /Users/dave/Library/Logs/DiagnosticReports/main_2017-06-20-201648-1_Daves-MacBook-Pro-2016.crash
```
This seems to show the PyInstall'ed app being killed by the sandbox because it tries to re-init itself...

**Comment by tallforasmurf** [comment]
Take a look at [the bootloader doc](http://pyinstaller.readthedocs.io/en/stable/advanced-topics.html#the-bootstrap-process-in-detail). Could it be the "launching itself as a child process" that is causing this?

**Comment by davegradwell** [comment]
Yep I think that's the cause, I'm just not sure what the solution is.
Maybe I should confer 'inherit' sandbox entitlements onto the bootloader's subprocess when codesigning (but I'm not sure exactly how).  Maybe there's an exception I can request in the entitlements.  It may be altogether impossible.  I certainly wouldn't expect the otherwise neat design of PyInstaller to be marred to work around the heinous Sandbox.
(Sidenote: I like your Sidetone app!)

**Comment by LoveDaisy** [comment]
hi @davegradwell I've run into a similar problem. When I codesign the standalone binary executable and tried to run it in terminal just as you did, I got the same error message 'illegal instruction'. Also I tried to codesign the bundle app and run it by directly double clicking, and nothing appeared. My bundle app just crashed without any signal.

I wonder if it is the same reason that caused the failures. Someone told me 'you use sandbox and you cannot run your app from terminal' (I cannot tell whether it is true). If it is true then the failure of bundle app could be caused by a different reason.

**Comment by tallforasmurf** [comment]
When a mac app launches and immediately disappears, there were probably console messages. Open Applications/Utilities/Console.app, clear the display, then double-click your app.

**Comment by davegradwell** [comment]
@LoveDaisy 
> I wonder if it is the same reason that caused the failures  

I would suspect so.

> 'you use sandbox and you cannot run your app from terminal'  

I do not find this to be true.

If you manage to crush sandbox into submission for a PyInstalled app I would be interested in hearing how.

**Comment by skiptomyliu** [comment]
@davegradwell , it's been a while but I was able to get pyinstaller sandbox working.  Submitted to the mac app store and was also approved.  I use the following script to sign my pyinstaller app:  
https://gist.github.com/skiptomyliu/d8e137b248b2d2d34ab3962b74229868#file-build-sh

Note that there is a child entitlement and parent entitlement.  The child entitlement is used on the pyinstaller app, which is launched as a child process from the parent application bundle.    I have a more detailed write up here (https://medium.com/python-pandemonium/embedding-a-python-application-in-macos-d866adfcaf94)

**Comment by davegradwell** [comment]
@skiptomyliu wow!  My situation (and possibly @LoveDaisy's) is slightly different in that it's the cardinal app itself which is Python, rather than an app intended for use as a child process.  Interesting to know that the Sandbox permits the child to 'recursively relaunch' (or however you want to envisage PyInstaller's behaviour) yet it kills the parent for doing this very same thing.  Also I found your blog a few months ago when I was working on this, and it was very insightful and helpful; appreciated.

**Comment by htgoebel** [comment]
If you develop a working version I'd appreciate a pull-request for either the manual or maybe adding some option or tool to PyInstaller.

**Comment by LoveDaisy** [comment]
@tallforasmurf Thanks for your suggestion. I checked my Console and found nothing when I was double clicking my app. Just nothing appeared, neither the app GUI nor anything in Console (just notice that I printed lots of logs in my app but none of them showed up). I wonder if something goes wrong before the app could start, say, it is prohibited by the system.

And one more thing. If I singed the executable only and started it from terminal, it just came up with "Illegal instruction: 4" just as Davegradwell has mentioned.

**Comment by KengoSawa2** [comment]
Thank you for everyone's information up to here.

I am also failing in sandboxing with the same problem.
When launching just one binary, a forbidden-sandbox-reinit error occurs.
I tried adding com.apple.security.inherit to Entitlements when code signing, but the problem was not resolved.
Is this a fundamental issue with the pyinstaller bootloader and can it be avoided?
If someone is working on this problem, I want information.

**Comment by BoboTiG** [comment]
Did you try that https://github.com/pyinstaller/pyinstaller/issues/4629#issuecomment-577438510? It works for me since a while.
